nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java

```diff
@@ -67,6 +67,10 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
             final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);
             final DataLakeFileClient fileClient = directoryClient.getFileClient(fileName);
 
+            if (fileClient.getProperties().isDirectory()) {
```

Comment:
I can see a possible optimization here.
According to @MuazmaZ's comment (https://github.com/apache/nifi/pull/4257#discussion_r423930566), `get***Client()` does not involve a network call but `getProperties()` does.
Most of the cases, this call would not be needed, because when the returned entity has content, it is definitely not a directory but a file. So I think this check could be moved after `session.write()` (line 74) and it needs to be checked only when the resulted flowfile is empty.
@pgyori, @MuazmaZ: what is your opinion?

Reply:
Unfortunately things get quite complicated in that case. If we call isDirectory() after session.write(), then the flowfile is already overwritten with the empty content, which means that after throwing the ProcessException, the flowfile that goes to the error output has no content (instead of the content of the original input flowfile). To avoid losing the content of the input flowfile, we would need to copy and store its content (before calling session.write()), and if isDirectory() returns true, we would need to load back this content to the original flowfile before throwing the exception. This would result in higher memory consumption in case of large input flowfiles.

Reply:
I agree @pgyori about the flow and for large flows that would be higher memory consumption. Also, a valid flowfile with empty content could be a real scenario where sometimes the process generates empty files to trigger other flows that I have seen with customers.

Reply:
@pgyori, @MuazmaZ Thanks for the clarification and the feedback. Then it is fine as it is now.
LGTM from my side.
@MuazmaZ Did you manage to make it work on your side or still NoSuchMethodError?

Reply:
@turcsanyip re-testing it right now. I will update shortly.

Reply:
Looks good to me after rebuild. +1

Failure to fetch file from Azure Data Lake Storage: org.apache.nifi.processor.exception.ProcessException: File Name (xyz) points to a directory. Full path: xyz/xyz

Reply:
Thanks, merging to master.